torchx/components/dist.py

```diff
@@ -80,6 +86,293 @@
+                ArgName(experiment_name='foo', run_name='bar')
+
+                >>> ArgName.parse_from(name="foo/", script="bar/baz.py")
+                ArgName(experiment_name='foo', run_name='baz')
+
+                >>> ArgName.parse_from(name="/bar", script="bar/baz.py")
+                ArgName(experiment_name=None, run_name='bar')
+
+                >>> ArgName.parse_from(name="foo/bar", m="foo.bar.baz")
+                ArgName(experiment_name='foo', run_name='bar')
+
+                >>> ArgName.parse_from(name="foo/", m="foo.bar.baz")
+                ArgName(experiment_name='foo', run_name='baz')
+
+                >>> ArgName.parse_from(name="/bar", m="foo.bar.baz")
+                ArgName(experiment_name=None, run_name='bar')
+
+                >>> ArgName.parse_from(name="foo/bar")
+                Traceback (most recent call last):
+                    ...
+                ValueError: No main module or script specified.
+
+                >>> ArgName.parse_from(name="foobar", m="foo.bar")
+                Traceback (most recent call last):
+                    ...
+                ValueError: Invalid name argument `foobar`...
+
+
+        Arguments:
+            name: ``{EXPERIMENT_NAME}/{RUN_NAME}``, ``/{RUN_NAME}``, or ``{EXPERIMENT_NAME}/``, or ``{RUN_NAME}``
+            m: the main module (e.g. ``foo.bar.baz`` for ``foo/bar/baz.py``)
+            script: path to the main script
+
+
+        Raises:
+            ValueError: if both ``m`` and ``script`` are empty or both are non-empty.
+            ValueError: if the ``name`` does not contain the experiment/run name delimiter: ``/``.
+        """
+        if not m and not script:
+            raise ValueError(
+                "No main module or script specified. Specify either a main module or a script path"
+            )
+        if m and script:
+            raise ValueError(
+                "Both main module and script set. Specify exactly one of: main module or script, but not both"
+            )
+
+        run_name = ""
+        experiment_name = ""
+
+        delim_idx = name.find("/")
+        if delim_idx < 0:
+            raise ValueError(
+                f"Invalid name argument `{name}`. It must contain the `/` delimiter and"
+                f" be of the form `FOO/BAR` or `FOO/` or `/BAR`"
+            )
+
+        # deal with /FOO (only run name), FOO/ (only exp name), FOO/BAR (both exp and run name)
+        if delim_idx >= 0 and delim_idx < len(name) - 1:
+            run_name = name[delim_idx + 1 :]
+        if delim_idx > 0:
+            experiment_name = name[:delim_idx]
+
+        if not run_name:
+            if m:  # use the last module name
+                run_name = m.rpartition(".")[2]
+            else:  # use script name w/ no extension
+                run_name = Path(safe_cast(str, script)).stem
+        return ArgName(experiment_name or None, run_name)
+
+
+@dataclass
+class ArgJ:
```

Comment:
make private?

also I don't love this name, _StructuredJ?

Reply:
renamed to `StructuredJArgument` and moved to module `torchx.components.structured_arg` so that we can reuse these higher-level arguments going forward to keep all the component arguments consistent.